File: cpp/tests/unit/test_new10_11_12_spec.cpp
// ARM SMMU v3 Conformance Gap Tests: NEW-10, NEW-11, NEW-12
// Copyright (c) 2024 John Greninger
//
// NEW-10 (§6.3.96): acknowledgeEventQueueOverflow() — OVACKFLG writable
// NEW-11 (§7.3.2):  reportUnsupportedTransaction() — F_UUT event generation
// NEW-12 (§3.9):    ATS Transaction Types (OT/TR/TT) enforcement

#include <gtest/gtest.h>
#include "smmu/smmu.h"
#include "smmu/types.h"
#include "smmu/address_space.h"
#include "smmu/configuration.h"
#include <memory>

using namespace smmu;

// ─────────────────────────────────────────────────────────────────────────────
// Shared helpers
// ─────────────────────────────────────────────────────────────────────────────

namespace {

static constexpr StreamID SID  = 0x20;
static constexpr PASID    PID  = 0;

// Enable SMMU + EVENTQEN
void enableSMMU(SMMU& smmu) {
    smmu.enable();
    smmu.setCR0(smmu.getCR0() | SMMU::CR0_EVENTQEN);
}

// Configure a minimal stage-1-only translation stream
void setupS1Stream(SMMU& smmu, StreamID sid = SID, PASID pasid = PID,
                   uint8_t eats = 0) {
    StreamConfig cfg;
    cfg.translationEnabled = true;
    cfg.stage1Enabled      = true;
    cfg.stage2Enabled      = false;
    cfg.bypassEnabled      = false;
    cfg.eats               = eats;
    ASSERT_TRUE(smmu.configureStream(sid, cfg).isOk());
    ASSERT_TRUE(smmu.enableStream(sid).isOk());
    ASSERT_TRUE(smmu.createStreamPASID(sid, pasid).isOk());
}

// Configure a bypass stream
void setupBypassStream(SMMU& smmu, StreamID sid = SID, PASID /*pasid*/ = PID) {
    StreamConfig cfg;
    cfg.translationEnabled = false;
    cfg.stage1Enabled      = false;
    cfg.stage2Enabled      = false;
    cfg.bypassEnabled      = true;
    cfg.eats               = 0;
    ASSERT_TRUE(smmu.configureStream(sid, cfg).isOk());
    ASSERT_TRUE(smmu.enableStream(sid).isOk());
}

// Map a page and return its IOVA
IOVA setupPage(SMMU& smmu, StreamID sid = SID, PASID pasid = PID,
               IOVA iova = 0x1000, PA pa = 0x2000) {
    PagePermissions perms;
    perms.read  = true;
    perms.write = true;
    perms.execute = false;
    EXPECT_TRUE(smmu.mapPage(sid, pasid, iova, pa, perms).isOk());
    return iova;
}

// Minimum queue size used by overflow tests so overflow is easily triggered.
static constexpr size_t SMALL_QUEUE = 16;

// Build an SMMU with a small event queue (makes overflow tests fast and controllable).
std::unique_ptr<SMMU> makeSmallQueueSMMU() {
    QueueConfiguration qcfg(SMALL_QUEUE, SMALL_QUEUE, SMALL_QUEUE);
    SMMUConfiguration cfg(qcfg, CacheConfiguration(), AddressConfiguration(), ResourceLimits());
    auto smmu = std::make_unique<SMMU>(cfg);
    smmu->enable();
    smmu->setCR0(smmu->getCR0() | SMMU::CR0_EVENTQEN);
    // Enable C_BAD_STREAMID recording so unknown-stream faults fill the queue.
    smmu->setCR2(SMMU::CR2_RECINVSID);
    return smmu;
}

// Fill the event queue to trigger overflow using a small-queue SMMU.
// The smmu must already have RECINVSID=1.
void fillEventQueue(SMMU& smmu, size_t maxSz) {
    // Fire (maxSz + 2) attempts to ensure overflow
    for (size_t i = 0; i <= maxSz + 1; ++i) {
        StreamID badSID = 0xDEAD0000u + static_cast<StreamID>(i);
        smmu.translate(badSID, 0, 0x1000, AccessType::Read);
    }
}

} // anonymous namespace

// =============================================================================
// NEW-10: acknowledgeEventQueueOverflow()
// =============================================================================

class New10EventQOverflow : public ::testing::Test {
protected:
    void SetUp() override {
        smmu_ = makeSmallQueueSMMU();
    }

    void TearDown() override {
        smmu_.reset();
    }

    std::unique_ptr<SMMU> smmu_;
};

// Test 1: acknowledge does NOT clear queued events; OVFLG becomes == OVACKFLG
TEST_F(New10EventQOverflow, AcknowledgeDoesNotClearQueue) {
    fillEventQueue(*smmu_, SMALL_QUEUE);

    // Verify overflow is active: OVFLG bit31 of PROD != OVACKFLG bit31 of CONS
    uint32_t prodBefore = smmu_->getEventqProdIndex();
    uint32_t consBefore = smmu_->getEventqConsIndex();
    bool overflowActive = ((prodBefore >> 31) & 1u) != ((consBefore >> 31) & 1u);
    EXPECT_TRUE(overflowActive) << "Expected overflow to be active before acknowledge";

    // How many events are queued
    size_t eventsBefore = smmu_->getEventQueue().size();
    EXPECT_GT(eventsBefore, 0u) << "Expected events in queue before acknowledge";

    // Acknowledge the overflow
    smmu_->acknowledgeEventQueueOverflow();

    // Events must still be in the queue (not cleared)
    size_t eventsAfter = smmu_->getEventQueue().size();
    EXPECT_EQ(eventsAfter, eventsBefore) << "acknowledgeEventQueueOverflow() must not clear events";

    // Overflow inactive: OVFLG == OVACKFLG after acknowledge
    uint32_t prodAfter = smmu_->getEventqProdIndex();
    uint32_t consAfter = smmu_->getEventqConsIndex();
    bool overflowInactive = ((prodAfter >> 31) & 1u) == ((consAfter >> 31) & 1u);
    EXPECT_TRUE(overflowInactive) << "Overflow should be inactive after acknowledge";
}

// Test 2: no-op when there is no overflow
TEST_F(New10EventQOverflow, NoOpWhenNoOverflow) {
    // No overflow has occurred yet
    uint32_t consBefore = smmu_->getEventqConsIndex();
    uint32_t ackFlagBefore = (consBefore >> 31) & 1u;

    smmu_->acknowledgeEventQueueOverflow();

    uint32_t consAfter = smmu_->getEventqConsIndex();
    uint32_t ackFlagAfter = (consAfter >> 31) & 1u;

    // OVACKFLG must be unchanged when no overflow was active
    EXPECT_EQ(ackFlagBefore, ackFlagAfter)
        << "acknowledgeEventQueueOverflow() must be no-op when no overflow";
}

// Test 3: acknowledge after draining events
TEST_F(New10EventQOverflow, AcknowledgeAfterDrain) {
    fillEventQueue(*smmu_, SMALL_QUEUE);

    // Overflow is active; drain all events
    smmu_->clearEventQueue();

    // After clear, prod/cons both reset to 0, so overflow flag is gone.
    // Calling acknowledge on a no-overflow state must still be a no-op.
    uint32_t consBefore = smmu_->getEventqConsIndex();
    uint32_t ackFlagBefore = (consBefore >> 31) & 1u;

    smmu_->acknowledgeEventQueueOverflow();

    uint32_t consAfter = smmu_->getEventqConsIndex();
    uint32_t ackFlagAfter = (consAfter >> 31) & 1u;

    // OVACKFLG unchanged (already matched OVFLG after reset)
    EXPECT_EQ(ackFlagBefore, ackFlagAfter);
    // Queue remains empty
    EXPECT_EQ(smmu_->getEventQueue().size(), 0u);
}

// =============================================================================
// NEW-11: reportUnsupportedTransaction() — F_UUT event
// =============================================================================

class New11FUut : public ::testing::Test {
protected:
    void SetUp() override {
        smmu_ = std::make_unique<SMMU>();
        enableSMMU(*smmu_);
    }

    void TearDown() override {
        smmu_.reset();
    }

    std::unique_ptr<SMMU> smmu_;
};

// Test 4: F_UUT event emitted when EVENTQEN=1
TEST_F(New11FUut, EmitsEventWhenEventqenEnabled) {
    smmu_->clearEventQueue();

    smmu_->reportUnsupportedTransaction(SID, PID, 0x5000,
                                        AccessType::Read,
                                        SecurityState::NonSecure);

    auto events = smmu_->getEventQueue();
    ASSERT_EQ(events.size(), 1u) << "Expected exactly one F_UUT event";
    EXPECT_EQ(events[0].type, EventType::F_UUT);
    EXPECT_EQ(events[0].streamID, SID);
    EXPECT_EQ(events[0].pasid, PID);
    EXPECT_EQ(events[0].address, static_cast<IOVA>(0x5000));
}

// Test 5: F_UUT silently dropped when EVENTQEN=0
TEST_F(New11FUut, SilentWhenEventqenDisabled) {
    // Disable EVENTQEN (keep SMMUEN=1)
    uint32_t cr0 = smmu_->getCR0();
    cr0 &= ~SMMU::CR0_EVENTQEN;
    smmu_->setCR0(cr0);

    smmu_->reportUnsupportedTransaction(SID, PID, 0x5000,
                                        AccessType::Read,
                                        SecurityState::NonSecure);

    EXPECT_EQ(smmu_->getEventQueue().size(), 0u)
        << "F_UUT must be silently dropped when EVENTQEN=0";
}

// Test 6: normal translation still works after reportUnsupportedTransaction()
TEST_F(New11FUut, TranslationStillWorksAfterFUut) {
    setupS1Stream(*smmu_);
    setupPage(*smmu_);

    // Emit an F_UUT for a different stream/address
    smmu_->reportUnsupportedTransaction(0x99, 0, 0xABCD,
                                        AccessType::Write,
                                        SecurityState::NonSecure);

    // Regular translation must succeed
    TranslationResult result = smmu_->translate(SID, PID, 0x1000, AccessType::Read);
    EXPECT_TRUE(result.isOk()) << "translate() must succeed after reportUnsupportedTransaction()";
}

// =============================================================================
// NEW-12: ATS Transaction Types (OT / TR / TT)
// =============================================================================

class New12Ats : public ::testing::Test {
protected:
    void SetUp() override {
        smmu_ = std::make_unique<SMMU>();
        enableSMMU(*smmu_);
    }

    void TearDown() override {
        smmu_.reset();
    }

    std::unique_ptr<SMMU> smmu_;
};

// Test 7: TR type with eats=0 → F_BAD_ATS_TREQ (§7.3.6)
TEST_F(New12Ats, TrType_EatsZero_BadAtsTreq) {
    setupS1Stream(*smmu_, SID, PID, /*eats=*/0);
    setupPage(*smmu_);
    smmu_->clearEventQueue();

    TranslationResult result = smmu_->translate(
        SID, PID, 0x1000, AccessType::Read,
        SecurityState::NonSecure,
        TransactionType::AtsTranslationRequest);

    EXPECT_FALSE(result.isOk()) << "TR on eats=0 stream must fail";

    auto events = smmu_->getEventQueue();
    ASSERT_FALSE(events.empty()) << "Expected F_BAD_ATS_TREQ event";
    EXPECT_EQ(events[0].type, EventType::F_BAD_ATS_TREQ);
}

// Test 8: TR type on bypass stream → F_BAD_ATS_TREQ (§7.3.6)
TEST_F(New12Ats, TrType_BypassStream_BadAtsTreq) {
    setupBypassStream(*smmu_);
    smmu_->clearEventQueue();

    TranslationResult result = smmu_->translate(
        SID, PID, 0x1000, AccessType::Read,
        SecurityState::NonSecure,
        TransactionType::AtsTranslationRequest);

    EXPECT_FALSE(result.isOk()) << "TR on bypass stream must fail";

    auto events = smmu_->getEventQueue();
    ASSERT_FALSE(events.empty()) << "Expected F_BAD_ATS_TREQ event";
    EXPECT_EQ(events[0].type, EventType::F_BAD_ATS_TREQ);
}

// Test 9: TR type with eats != 0 and translation-enabled stream → success
TEST_F(New12Ats, TrType_EatsNonZero_Succeeds) {
    setupS1Stream(*smmu_, SID, PID, /*eats=*/2);
    setupPage(*smmu_);
    smmu_->clearEventQueue();

    TranslationResult result = smmu_->translate(
        SID, PID, 0x1000, AccessType::Read,
        SecurityState::NonSecure,
        TransactionType::AtsTranslationRequest);

    EXPECT_TRUE(result.isOk()) << "TR with eats=2 on translation-enabled stream must succeed";
    EXPECT_TRUE(smmu_->getEventQueue().empty()) << "No error event expected on success";
}

// Test 10: TT type with ATSCHK=0 → no re-check, succeeds
TEST_F(New12Ats, TtType_AtschkOff_NoRecheck) {
    setupS1Stream(*smmu_);
    setupPage(*smmu_);
    // Ensure ATSCHK=0
    smmu_->setCR0(smmu_->getCR0() & ~SMMU::CR0_ATSCHK);
    smmu_->clearEventQueue();

    TranslationResult result = smmu_->translate(
        SID, PID, 0x1000, AccessType::Read,
        SecurityState::NonSecure,
        TransactionType::AtsTranslated);

    EXPECT_TRUE(result.isOk()) << "TT with ATSCHK=0 must succeed without re-check";
    EXPECT_TRUE(smmu_->getEventQueue().empty()) << "No event expected with ATSCHK=0";
}

// Test 11: TT type with ATSCHK=1 and mapped page → success
TEST_F(New12Ats, TtType_AtschkOn_MappedPage_Succeeds) {
    setupS1Stream(*smmu_);
    setupPage(*smmu_);
    smmu_->setCR0(smmu_->getCR0() | SMMU::CR0_ATSCHK);
    smmu_->clearEventQueue();

    TranslationResult result = smmu_->translate(
        SID, PID, 0x1000, AccessType::Read,
        SecurityState::NonSecure,
        TransactionType::AtsTranslated);

    EXPECT_TRUE(result.isOk()) << "TT with ATSCHK=1 on mapped page must succeed";
    EXPECT_TRUE(smmu_->getEventQueue().empty()) << "No error event expected on success";
}

// Test 12: TT type with ATSCHK=1 and unmapped address → F_TRANSL_FORBIDDEN (§7.3.8)
TEST_F(New12Ats, TtType_AtschkOn_UnmappedPage_TranslForbidden) {
    setupS1Stream(*smmu_);
    // Do NOT map any page at 0x9000
    smmu_->setCR0(smmu_->getCR0() | SMMU::CR0_ATSCHK);
    smmu_->clearEventQueue();

    TranslationResult result = smmu_->translate(
        SID, PID, 0x9000, AccessType::Read,
        SecurityState::NonSecure,
        TransactionType::AtsTranslated);

    EXPECT_FALSE(result.isOk()) << "TT with ATSCHK=1 on unmapped page must fail";

    auto events = smmu_->getEventQueue();
    ASSERT_FALSE(events.empty()) << "Expected F_TRANSL_FORBIDDEN event";
    EXPECT_EQ(events[0].type, EventType::F_TRANSL_FORBIDDEN);
}

// Test 13: Ordinary type (default) — behavior unchanged
TEST_F(New12Ats, OrdinaryType_BehaviorPreserved) {
    setupS1Stream(*smmu_);
    setupPage(*smmu_);
    smmu_->clearEventQueue();

    // With explicit Ordinary
    TranslationResult r1 = smmu_->translate(
        SID, PID, 0x1000, AccessType::Read,
        SecurityState::NonSecure,
        TransactionType::Ordinary);
    EXPECT_TRUE(r1.isOk()) << "Ordinary type on mapped page must succeed";

    // With default (no transactionType argument) — same address, same PASID
    // Need to re-map because the translate() call may exhaust nothing here;
    // just verify the default overload still compiles and works.
    TranslationResult r2 = smmu_->translate(SID, PID, 0x1000, AccessType::Read);
    EXPECT_TRUE(r2.isOk()) << "Default (Ordinary) type must also succeed";

    EXPECT_TRUE(smmu_->getEventQueue().empty()) << "No error events for ordinary type";
}

File: rust/smmu/tests/test_new10_11_12_spec.rs
//! Tests for NEW-10, NEW-11, and NEW-12 conformance gaps.
//!
//! NEW-10 (§6.3.96): `acknowledge_eventq_overflow()` — copies OVFLG into OVACKFLG
//!                   without clearing queue contents.
//! NEW-11 (§7.3.2):  `report_unsupported_transaction()` — generates `F_UUT` event,
//!                   gated on `CR0.EVENTQEN`.
//! NEW-12 (§3.9):    `TransactionType` / `translate_with_type()` — ATS translation
//!                   request and translated-transaction checks.

use smmu::types::{
    AccessType, EventEntry, EventType, IOVA, PASID, PagePermissions, SecurityState, StreamConfig,
    StreamID,
};
use smmu::{TransactionType, SMMU};

// ── NEW-10 tests ─────────────────────────────────────────────────────────────

/// NEW-10: `acknowledge_eventq_overflow()` must copy OVFLG into OVACKFLG
/// without discarding queued events.
#[test]
fn test_new10_acknowledge_overflow_does_not_clear_queue() {
    // Use smallest allowed event queue so we can overflow quickly.
    use smmu::types::{QueueConfig, SMMUConfig};
    let capacity: usize = 16;
    let smmu = SMMU::with_config(SMMUConfig::from(QueueConfig {
        event_queue_size: capacity,
        command_queue_size: 32,
        pri_queue_size: 32,
    }));
    // Enable SMMU + EVENTQEN so events are accepted.
    smmu.set_cr0(SMMU::CR0_SMMUEN | SMMU::CR0_EVENTQEN | SMMU::CR0_CMDQEN);
    smmu.set_cr2(SMMU::CR2_RECINVSID);

    // Fill queue to capacity.
    for i in 0..u32::try_from(capacity).unwrap() {
        let ev = EventEntry {
            event_type: EventType::FTranslation,
            stream_id: i,
            ..EventEntry::zeroed()
        };
        let _ = smmu.submit_event(ev);
    }

    // Trigger an overflow by trying to add one more via translation fault.
    let sid = StreamID::new(0x99).unwrap();
    let pasid = PASID::new(0).unwrap();
    let iova = IOVA::new(0x1000).unwrap();
    let _ = smmu.translate(sid, pasid, iova, AccessType::Read, SecurityState::NonSecure);

    // OVFLG must be active: prod[31] != cons[31].
    let prod_before = smmu.get_eventq_prod();
    let cons_before = smmu.eventq_cons_index();
    assert_ne!(
        (prod_before >> 31) & 1,
        (cons_before >> 31) & 1,
        "OVFLG must be active after overflow"
    );

    // Call the method under test.
    smmu.acknowledge_eventq_overflow();

    // (a) Queue contents must be unchanged — all capacity events still present.
    let events = smmu.get_events();
    assert_eq!(
        events.len(),
        capacity,
        "Queue must retain all {capacity} events after acknowledge (no clear)"
    );

    // (b) OVFLG and OVACKFLG must now be equal (overflow acknowledged).
    let prod_after = smmu.get_eventq_prod();
    let cons_after = smmu.eventq_cons_index();
    assert_eq!(
        (prod_after >> 31) & 1,
        (cons_after >> 31) & 1,
        "OVACKFLG must equal OVFLG after acknowledge_eventq_overflow()"
    );
}

/// NEW-10: When no overflow is present, `acknowledge_eventq_overflow()` must be
/// a no-op (both OVFLG and OVACKFLG remain 0).
#[test]
fn test_new10_acknowledge_no_op_when_no_overflow() {
    let smmu = SMMU::new();
    smmu.set_cr0(SMMU::CR0_SMMUEN | SMMU::CR0_EVENTQEN | SMMU::CR0_CMDQEN);

    // No overflow has occurred — both bits are 0.
    let prod_before = smmu.get_eventq_prod();
    let cons_before = smmu.eventq_cons_index();
    assert_eq!((prod_before >> 31) & 1, 0, "OVFLG must start at 0");
    assert_eq!((cons_before >> 31) & 1, 0, "OVACKFLG must start at 0");

    smmu.acknowledge_eventq_overflow();

    let prod_after = smmu.get_eventq_prod();
    let cons_after = smmu.eventq_cons_index();
    assert_eq!((prod_after >> 31) & 1, 0, "OVFLG must remain 0 (no-op)");
    assert_eq!((cons_after >> 31) & 1, 0, "OVACKFLG must remain 0 (no-op)");
}

// ── NEW-11 tests ──────────────────────────────────────────────────────────────

/// NEW-11: `report_unsupported_transaction()` must emit an `FUut` event when
/// `CR0.EVENTQEN` is set.
#[test]
fn test_new11_fuut_emitted_when_eventqen() {
    let smmu = SMMU::new();
    // Enable SMMU and EVENTQEN.
    smmu.set_cr0(SMMU::CR0_SMMUEN | SMMU::CR0_EVENTQEN | SMMU::CR0_CMDQEN);

    let sid = StreamID::new(5).unwrap();
    let pasid = PASID::new(0).unwrap();
    let iova = IOVA::new(0x2000).unwrap();

    let result = smmu.report_unsupported_transaction(
        sid,
        pasid,
        iova,
        AccessType::Read,
        SecurityState::NonSecure,
    );
    assert!(result.is_ok(), "report_unsupported_transaction() must succeed");

    let events = smmu.get_events();
    assert_eq!(events.len(), 1, "Exactly one F_UUT event must be queued");
    let ev = &events[0];
    assert_eq!(ev.event_type, EventType::FUut, "Event type must be FUut");
    assert_eq!(ev.stream_id, 5, "stream_id must match");
    assert_eq!(ev.pasid, 0, "pasid must match");
    assert_eq!(ev.address, 0x2000, "address must match");
    assert_eq!(ev.security_state, SecurityState::NonSecure, "security_state must match");
    assert_eq!(ev.event_class, 2, "event_class must be 2 (IN) per NEW-1");
}

/// NEW-11: `report_unsupported_transaction()` must silently drop the event when
/// `CR0.EVENTQEN` is not set.
#[test]
fn test_new11_fuut_silent_when_eventqen_disabled() {
    let smmu = SMMU::new();
    // EVENTQEN is NOT set.
    smmu.set_cr0(SMMU::CR0_SMMUEN | SMMU::CR0_CMDQEN);

    let sid = StreamID::new(7).unwrap();
    let pasid = PASID::new(0).unwrap();
    let iova = IOVA::new(0x3000).unwrap();

    let result = smmu.report_unsupported_transaction(
        sid,
        pasid,
        iova,
        AccessType::Write,
        SecurityState::NonSecure,
    );
    assert!(result.is_ok(), "report_unsupported_transaction() must still return Ok");

    let events = smmu.get_events();
    assert!(events.is_empty(), "No event must be queued when EVENTQEN=0");
}

// ── NEW-12 tests ──────────────────────────────────────────────────────────────

/// Helper: create SMMU with a stage-1 stream on StreamID 1 with eats=0 (no ATS).
fn make_smmu_s1_no_ats() -> SMMU {
    let smmu = SMMU::new();
    smmu.set_cr0(SMMU::CR0_SMMUEN | SMMU::CR0_EVENTQEN | SMMU::CR0_CMDQEN);
    let sid = StreamID::new(1).unwrap();
    let cfg = StreamConfig {
        eats: 0,
        t0sz: 0, // no VA-range restriction
        ips: 6,  // 52-bit IPS
        ..StreamConfig::stage1_only()
    };
    smmu.configure_stream(sid, cfg).unwrap();
    let pasid = PASID::new(0).unwrap();
    smmu.create_pasid(sid, pasid).unwrap();
    // Map a page for StreamID=1, PASID=0, IOVA=0x1000.
    let iova = IOVA::new(0x1000).unwrap();
    smmu.map_page(
        sid,
        pasid,
        iova,
        smmu::types::PA::new(0x8000_1000).unwrap(),
        PagePermissions::read_write(),
        SecurityState::NonSecure,
    )
    .unwrap();
    smmu
}

/// Helper: create SMMU with a stage-1 stream on StreamID 1 with eats=1 (ATS enabled).
fn make_smmu_s1_with_ats() -> SMMU {
    let smmu = SMMU::new();
    smmu.set_cr0(SMMU::CR0_SMMUEN | SMMU::CR0_EVENTQEN | SMMU::CR0_CMDQEN);
    let sid = StreamID::new(1).unwrap();
    let cfg = StreamConfig {
        eats: 1,
        t0sz: 0,
        ips: 6,
        ..StreamConfig::stage1_only()
    };
    smmu.configure_stream(sid, cfg).unwrap();
    let pasid = PASID::new(0).unwrap();
    smmu.create_pasid(sid, pasid).unwrap();
    let iova = IOVA::new(0x1000).unwrap();
    smmu.map_page(
        sid,
        pasid,
        iova,
        smmu::types::PA::new(0x8000_1000).unwrap(),
        PagePermissions::read_write(),
        SecurityState::NonSecure,
    )
    .unwrap();
    smmu
}

/// NEW-12: ATS Translation Request (TR) on a stream with `eats=0` must generate
/// `F_BAD_ATS_TREQ` (§7.3.6 event 0x05) and return an error.
#[test]
fn test_new12_tr_eats_zero_bad_ats_treq() {
    let smmu = make_smmu_s1_no_ats();
    let sid = StreamID::new(1).unwrap();
    let pasid = PASID::new(0).unwrap();
    let iova = IOVA::new(0x1000).unwrap();

    let result = smmu.translate_with_type(
        sid,
        pasid,
        iova,
        AccessType::Read,
        SecurityState::NonSecure,
        TransactionType::AtsTranslationRequest,
    );
    assert!(result.is_err(), "TR on eats=0 stream must fail");

    let events = smmu.get_events();
    let bad_ats_treq_count =
        events.iter().filter(|e| e.event_type == EventType::FBadAtsTreq).count();
    assert_eq!(
        bad_ats_treq_count,
        1,
        "Exactly one F_BAD_ATS_TREQ event must be emitted; event types: {:?}",
        events.iter().map(|e| e.event_type).collect::<Vec<_>>()
    );
}

/// NEW-12: ATS Translation Request (TR) on a bypass stream must also generate
/// `F_BAD_ATS_TREQ` (§7.3.6 event 0x05) — bypass streams do not support ATS.
#[test]
fn test_new12_tr_bypass_stream_bad_ats_treq() {
    let smmu = SMMU::new();
    smmu.set_cr0(SMMU::CR0_SMMUEN | SMMU::CR0_EVENTQEN | SMMU::CR0_CMDQEN);
    let sid = StreamID::new(2).unwrap();
    let cfg = StreamConfig { eats: 1, ..StreamConfig::bypass() };
    smmu.configure_stream(sid, cfg).unwrap();

    let pasid = PASID::new(0).unwrap();
    let iova = IOVA::new(0x1000).unwrap();

    let result = smmu.translate_with_type(
        sid,
        pasid,
        iova,
        AccessType::Read,
        SecurityState::NonSecure,
        TransactionType::AtsTranslationRequest,
    );
    assert!(result.is_err(), "TR on bypass stream must fail");

    let events = smmu.get_events();
    let found = events.iter().any(|e| e.event_type == EventType::FBadAtsTreq);
    assert!(found, "F_BAD_ATS_TREQ must be emitted for TR on bypass stream");
}

/// NEW-12: ATS Translation Request (TR) on a stream with eats!=0 and translation
/// enabled must succeed and return a translated PA.
#[test]
fn test_new12_tr_eats_nonzero_succeeds() {
    let smmu = make_smmu_s1_with_ats();
    let sid = StreamID::new(1).unwrap();
    let pasid = PASID::new(0).unwrap();
    let iova = IOVA::new(0x1000).unwrap();

    let result = smmu.translate_with_type(
        sid,
        pasid,
        iova,
        AccessType::Read,
        SecurityState::NonSecure,
        TransactionType::AtsTranslationRequest,
    );
    assert!(result.is_ok(), "TR on eats!=0 stream with valid mapping must succeed; got {result:?}");
}

/// NEW-12: ATS Translated transaction (TT) with ATSCHK=0 must succeed without
/// re-checking the mapping (even if IOVA is not actually mapped).
#[test]
fn test_new12_tt_atschk_off_no_recheck() {
    let smmu = make_smmu_s1_with_ats();
    let sid = StreamID::new(1).unwrap();
    let pasid = PASID::new(0).unwrap();
    // Use the mapped IOVA — ATSCHK=0 so no extra F_BAD_ATS_TREQ re-check step.
    let iova = IOVA::new(0x1000).unwrap();

    // ATSCHK is bit 4 of CR0 — we keep it 0 (default).
    let result = smmu.translate_with_type(
        sid,
        pasid,
        iova,
        AccessType::Read,
        SecurityState::NonSecure,
        TransactionType::AtsTranslated,
    );
    // Without ATSCHK, the TT request uses the ordinary translation path;
    // the mapped page means the translation succeeds with no extra event.
    assert!(result.is_ok(), "TT with ATSCHK=0 on mapped page must succeed");
    // No ATS error events must be emitted (ATSCHK=0, no re-check).
    let events = smmu.get_events();
    let bad_treq = events.iter().any(|e| e.event_type == EventType::FBadAtsTreq);
    let transl_forbidden = events.iter().any(|e| e.event_type == EventType::FTranslForbidden);
    assert!(!bad_treq, "F_BAD_ATS_TREQ must NOT be emitted when ATSCHK=0");
    assert!(!transl_forbidden, "F_TRANSL_FORBIDDEN must NOT be emitted when ATSCHK=0");
}

/// NEW-12: ATS Translated transaction (TT) with ATSCHK=1 and a mapped IOVA
/// must succeed (re-check passes).
#[test]
fn test_new12_tt_atschk_on_mapped_succeeds() {
    let smmu = make_smmu_s1_with_ats();
    // Set ATSCHK=1.
    smmu.set_cr0(
        SMMU::CR0_SMMUEN | SMMU::CR0_EVENTQEN | SMMU::CR0_CMDQEN | SMMU::CR0_ATSCHK,
    );
    let sid = StreamID::new(1).unwrap();
    let pasid = PASID::new(0).unwrap();
    let iova = IOVA::new(0x1000).unwrap(); // this page IS mapped

    let result = smmu.translate_with_type(
        sid,
        pasid,
        iova,
        AccessType::Read,
        SecurityState::NonSecure,
        TransactionType::AtsTranslated,
    );
    assert!(result.is_ok(), "TT with ATSCHK=1 and valid mapping must succeed; got {result:?}");

    let events = smmu.get_events();
    let transl_forbidden = events.iter().any(|e| e.event_type == EventType::FTranslForbidden);
    assert!(!transl_forbidden, "F_TRANSL_FORBIDDEN must NOT be emitted for a valid TT");
}

/// NEW-12: ATS Translated transaction (TT) with `ATSCHK=1` and an UNMAPPED IOVA
/// must generate `F_TRANSL_FORBIDDEN` (§7.3.8 event 0x07) and return an error.
#[test]
fn test_new12_tt_atschk_on_unmapped_transl_forbidden() {
    let smmu = make_smmu_s1_with_ats();
    // Set ATSCHK=1.
    smmu.set_cr0(
        SMMU::CR0_SMMUEN | SMMU::CR0_EVENTQEN | SMMU::CR0_CMDQEN | SMMU::CR0_ATSCHK,
    );
    let sid = StreamID::new(1).unwrap();
    let pasid = PASID::new(0).unwrap();
    let iova = IOVA::new(0x9000).unwrap(); // NOT mapped

    let result = smmu.translate_with_type(
        sid,
        pasid,
        iova,
        AccessType::Read,
        SecurityState::NonSecure,
        TransactionType::AtsTranslated,
    );
    assert!(result.is_err(), "TT with ATSCHK=1 and unmapped IOVA must fail");

    let events = smmu.get_events();
    let transl_forbidden_count =
        events.iter().filter(|e| e.event_type == EventType::FTranslForbidden).count();
    assert_eq!(
        transl_forbidden_count,
        1,
        "Exactly one F_TRANSL_FORBIDDEN event must be emitted; event types: {:?}",
        events.iter().map(|e| e.event_type).collect::<Vec<_>>()
    );
}

/// NEW-12: Ordinary transactions must be completely unaffected by the new
/// `translate_with_type()` infrastructure — existing behavior preserved.
#[test]
fn test_new12_ordinary_behavior_preserved() {
    let smmu = make_smmu_s1_with_ats();
    let sid = StreamID::new(1).unwrap();
    let pasid = PASID::new(0).unwrap();
    let iova = IOVA::new(0x1000).unwrap();

    // translate() delegates to translate_with_type(..., Ordinary).
    let result_via_translate = smmu.translate(
        sid,
        pasid,
        iova,
        AccessType::Read,
        SecurityState::NonSecure,
    );
    smmu.clear_event_queue();

    let result_via_type = smmu.translate_with_type(
        sid,
        pasid,
        iova,
        AccessType::Read,
        SecurityState::NonSecure,
        TransactionType::Ordinary,
    );

    assert_eq!(
        result_via_translate.is_ok(),
        result_via_type.is_ok(),
        "Ordinary translate_with_type must match translate() success/failure"
    );
}
